onnxruntime/core/session/inference_session.cc

```diff
@@ -984,14 +984,19 @@ common::Status InferenceSession::Load() {
 
 common::Status InferenceSession::TransformGraph(onnxruntime::Graph& graph, bool saving_model_in_ort_format) {
   // The transformer order:
-  // 1. ensure potential QDQ node units have unique DQ nodes (required transformer).
+  // 1. Ensure we inline as many functions as possible. We refer to it as Ahead Of Time (AOT) function inlining.
+  // 2. ensure potential QDQ node units have unique DQ nodes (required transformer).
   //    - This is a required transformer as the ORT code has a hard requirement there are no overlapping QDQ node units.
   //    - We run it here in case optimizers are disabled.
-  // 2. run level 1 optimizations. these only use ONNX operators.
-  // 3. partition nodes based on EP capabilities. EPs may fuse nodes during this process.
-  // 4. run level 2+ optimizations. level 2 and 3 optimizations use contrib ops.
-  // 5. insert cast nodes (required transformer).
-  // 6. insert copy nodes (required transformer).
+  // 3. run level 1 optimizations. these only use ONNX operators.
+  // 4. partition nodes based on EP capabilities. EPs may fuse nodes during this process.
+  // 5. run level 2+ optimizations. level 2 and 3 optimizations use contrib ops.
+  // 6. insert cast nodes (required transformer).
+  // 7. insert copy nodes (required transformer).
+
+  // Run Ahead Of time function inlining
+  GraphPartitioner partitioner(kernel_registry_manager_, execution_providers_);
+  ORT_RETURN_IF_ERROR_SESSIONID_(partitioner.IntelligentAotFunctionInlining(graph, execution_providers_, kernel_registry_manager_));
```

Comment:
I don't think we need to invoke this for all graphs. For graphs that don't have model local functions (the ones exported by torch script and not Dynamo), can we skip this? Want to make sure there's absolutely no change in behavior for older exported models.

Reply:
One of the issues with the current implementation is that a schema based functions such as `CastLike` are not being inlined and then constant folded. That results in a lot of nodes downstream of them not constant folded, and this is one of the complaints. So we want this change and for that we need to examine the graphs.